## Task 1

In [1]:
def check_board(alice_nums, bob_nums, row, col):
    # Check if binary sum of Alice's numbers == 0
    if (sum(alice_nums) % 2) != 0:
        return False
    
    # Check if binary sum of Bob's numbers == 1
    if (sum(bob_nums) % 2) != 1:
        return False
    
    # Check if Alice and Bob placed same number in cell where her row and his column intersect
    if alice_nums[col] != bob_nums[row]:
        return False
    
    return True


# print(check_board([1, 0, 1], [1, 1, 0], 1, 2))
# print(check_board([0, 0, 0], [0, 0, 1], 2, 2))
# print(check_board([0, 1, 1], [1, 0, 0], 0, 1))

In [2]:
import random
import numpy as np


def simulate_task1():
    '''
    returns True if Alice and Bob won
            False if Alice and Bob lost
    '''
    
    
    # Richard randomly chooses row and column
    row = random.randint(0, 2)
    col = random.randint(0, 2)
    
    # the values filled are represented by the binary representation of the random int
    rand1 = random.randint(0, 7)
    rand2 = random.randint(0, 7)
    
    alice_nums = [rand1 & (1 << x) for x in range(3)]
    bob_nums = [rand2 & (1 << x) for x in range(3)]
    
    return check_board(alice_nums, bob_nums, row, col)


results = [simulate_task1() for _ in range(500000)]

print(f'{sum(results)} games won')
print(f'{len(results)} games played')
print(f'{sum(results)/len(results):.5f} games won')

34895 games won
500000 games played
0.06979 games won


In [3]:
def generate_all_boards_task1():    
    all_boards = []
    for row in range(3):
        for col in range(3):
            for rand1 in range(8):
                for rand2 in range(8):
                    alice_nums = [rand1 & (1 << x) for x in range(3)]
                    bob_nums = [rand2 & (1 << x) for x in range(3)]
                    all_boards.append((alice_nums, bob_nums, row, col))
    return all_boards

results = [check_board(*board) for board in generate_all_boards_task1()]

print(f'{sum(results)} winning boards')
print(f'{len(results)} total possible boards')
print(f'{sum(results)/len(results):.5f} probability of winning')

40 winning boards
576 total possible boards
0.06944 probability of winning


Through simulation of all possible states of the board, we can determine that the probability of winning would be $\frac{40}{576}=\frac{5}{72}\approx0.06944$. This is rather close to the value obtained from above.

## Task 2

We make a minor adjustment to the solution for Task 1. Instead of randomly generating all 3 bits, we only generate the first 2 bits. The first 2 bits will determine what the last bit is in order to satisfy the binary sum condition.

Alice:

| Bit 1 | Bit 2 | Bit 3 |
|:-:|:-:|:-:|
| 0 | 0 | 0 |
| 0 | 1 | 1 |
| 1 | 0 | 1 |
| 1 | 1 | 0 |

Bob:

| Bit 1 | Bit 2 | Bit 3 |
|:-:|:-:|:-:|
| 0 | 0 | 1 |
| 0 | 1 | 0 |
| 1 | 0 | 0 |
| 1 | 1 | 1 |

In [4]:
def simulate_task2():
    '''
    returns True if Alice and Bob won
            False if Alice and Bob lost
    '''
    
    # Mostly similar to Task 1
    
    row = random.randint(0, 2)
    col = random.randint(0, 2)
    
    # Only generate first 2 bits
    # Last bit will be uniquely determined by the first 2 bits
    rand1 = random.randint(0, 3)
    rand2 = random.randint(0, 3)
    
    alice_nums = [rand1 & (1 << x) for x in range(2)]
    alice_nums.append((2 - sum(alice_nums) % 2))
    bob_nums = [rand2 & (1 << x) for x in range(2)]
    bob_nums.append((3 - sum(bob_nums) % 2))
    
    return check_board(alice_nums, bob_nums, row, col)


results = [simulate_task2() for _ in range(500000)]

print('Task 2')
print(f'{sum(results)} games won')
print(f'{len(results)} games played')
print(f'{sum(results)/len(results):.5f} games won')

Task 2
138956 games won
500000 games played
0.27791 games won


In [5]:
def generate_all_boards_task2():    
    all_boards = []
    for row in range(3):
        for col in range(3):
            for rand1 in range(4):
                for rand2 in range(4):
                    alice_nums = [rand1 & (1 << x) for x in range(2)]
                    alice_nums.append((2 - sum(alice_nums) % 2))
                    bob_nums = [rand2 & (1 << x) for x in range(2)]
                    bob_nums.append((3 - sum(bob_nums) % 2))
                    all_boards.append((alice_nums, bob_nums, row, col))
    return all_boards

results = [check_board(*board) for board in generate_all_boards_task2()]

print(f'{sum(results)} winning boards')
print(f'{len(results)} total possible boards')
print(f'{sum(results)/len(results):.5f} probability of winning')

40 winning boards
144 total possible boards
0.27778 probability of winning
